jgrapht-core/src/main/java/org/jgrapht/alg/CliqueMinimalSeparatorDecomposition.java

```diff
@@ -0,0 +1,452 @@
+    {
+        if (chordalGraph == null)
+            computeMinimalTriangulation();
+
+        separators = new HashSet<Set<V>>();
+
+        // initialize g' as subgraph of graph (same vertices and edges)
+        @SuppressWarnings("unchecked")
+        SimpleGraph<V, E> gprime = (SimpleGraph<V, E>) graph.clone();
+
+        // initialize h' as subgraph of chordalGraph (same vertices and edges)
+        @SuppressWarnings("unchecked")
+        SimpleGraph<V, E> hprime = (SimpleGraph<V, E>) chordalGraph.clone();
+
+        atoms = new HashSet<Set<V>>();
+
+        Iterator<V> iterator = meo.descendingIterator();
+        while (iterator.hasNext()) {
+            V v = iterator.next();
+            if (generators.contains(v)) {
+                Set<V> separator = new HashSet<V>(Graphs.neighborListOf(
+                    hprime,
+                    v));
+
+                if (isClique(graph, separator)) {
+                    if (separator.size() > 0) {
+                        if (separators.contains(separator)) {
+                            fullComponentCount.put(
+                                separator,
+                                fullComponentCount.get(separator) + 1);
+                        } else {
+                            fullComponentCount.put(separator, 2);
+                            separators.add(separator);
+                        }
+                    }
+                    @SuppressWarnings("unchecked")
+                    SimpleGraph<V, E> tmpGraph = 
+                        (SimpleGraph<V, E>) gprime.clone();
+
+                    tmpGraph.removeAllVertices(separator);
+                    ConnectivityInspector<V, E> con = 
+                        new ConnectivityInspector<V, E>(tmpGraph);
+                    if (con.isGraphConnected()) {
+                        throw new RuntimeException(
+                            "separator did not separate the graph");
+                    }
+                    for (Set<V> component : con.connectedSets()) {
+                        if (component.contains(v)) {
+                            gprime.removeAllVertices(component);
+                            component.addAll(separator);
+                            atoms.add(new HashSet<V>(component));
+                            assert (component.size() > 0);
+                            break;
+                        }
+                    }
+                }
+            }
+
+            hprime.removeVertex(v);
+        }
+
+        if (gprime.vertexSet().size() > 0) {
+            atoms.add(new HashSet<V>(gprime.vertexSet()));
+        }
+    }
+
+    /**
+     * Check whether the subgraph of <code>graph</code> induced by the given
+     * <code>vertices</code> is complete, i.e. a clique.
+     * 
+     * @param graph the graph.
+     * @param vertices the vertices to induce the subgraph from.
+     * @return true if the induced subgraph is a clique.
+     */
+    public static <V, E> boolean isClique(

```

Comment:
This should be private; later we can move it to a more appropriate general utility class.